libs/parallelism/algorithms/include/hpx/parallel/util/loop.hpp

```diff
@@ -297,66 +352,258 @@ namespace hpx { namespace parallel { namespace util {
         }
     }    // namespace detail
 
-    template <typename ExPolicy, typename Iter, typename F>
-    HPX_HOST_DEVICE HPX_FORCEINLINE constexpr typename std::enable_if<
-        !hpx::is_vectorpack_execution_policy<ExPolicy>::value, Iter>::type
-    loop_n(Iter it, std::size_t count, F&& f)
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename ExPolicy>
+    struct loop_n_t final : hpx::functional::tag_fallback<loop_n_t<ExPolicy>>
     {
-        using pred = std::integral_constant<bool,
-            hpx::traits::is_random_access_iterator<Iter>::value ||
-                std::is_integral<Iter>::value>;
+    private:
+        template <typename Iter, typename F>
+        friend HPX_HOST_DEVICE HPX_FORCEINLINE constexpr Iter
+        tag_fallback_invoke(hpx::parallel::util::loop_n_t<ExPolicy>, Iter it,
+            std::size_t count, F&& f)
+        {
+            using pred = std::integral_constant<bool,
+                hpx::traits::is_random_access_iterator<Iter>::value ||
+                    std::is_integral<Iter>::value>;
+
+            return detail::loop_n<Iter>::call(
+                it, count, std::forward<F>(f), pred());
+        }
+
+        template <typename Iter, typename CancelToken, typename F>
+        friend HPX_HOST_DEVICE HPX_FORCEINLINE constexpr Iter
+        tag_fallback_invoke(hpx::parallel::util::loop_n_t<ExPolicy>, Iter it,
+            std::size_t count, CancelToken& tok, F&& f)
+        {
+            using pred = std::integral_constant<bool,
+                hpx::traits::is_random_access_iterator<Iter>::value ||
+                    std::is_integral<Iter>::value>;
+
+            return detail::loop_n<Iter>::call(
+                it, count, tok, std::forward<F>(f), pred());
+        }
+    };
 
-        return detail::loop_n<Iter>::call(
-            it, count, std::forward<F>(f), pred());
+#if !defined(HPX_COMPUTE_DEVICE_CODE)
+    template <typename ExPolicy>
+    HPX_INLINE_CONSTEXPR_VARIABLE loop_n_t<ExPolicy> loop_n =
+        loop_n_t<ExPolicy>{};
```

Comment:
@hkaiser this was the CPO I was talking about that is not in detail. I don't think it needs to be in a non-detail namespace.

Reply:
This has been moved now. Please review.